yggdrasil: init at 0.3.8 and add service module

gazally opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Use an encrypted ipv6 overlay network.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
See also #63855 by @Lassulus, which has a different derivation for `yggdrasil` which also works with the service module.  This is my first attempt to write a Go derivation so I'll be happy to rebase this pull request on #63855 if that is preferable.

gazally committed:
maintainers: add gazally

gazally committed:
yggdrasil: init at 0.3.8

gazally committed:
nixos/yggdrasil: add service

Lassulus commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -0,0 +1,119 @@
+      };
+    };
+
+    networking.firewall.allowedTCPPorts = mkIf (cfg.port != null) [ cfg.port ];
```
https://nixos.org/nixpkgs/manual/#reviewing-contributions-new-modules
`For example, enabling a module should not open firewall ports by default. `

gazally replied:
I don't understand what you're asking for here.  The default for `cfg.port` is `null` so a port only gets opened if you set it to something else.

Lassulus replied:
But why? maybe people want to use in on another port and require special firewall rules? And why only open it, if it's a non default port?

gazally replied:
Just as a convenience to keep everything related to yggdrasil configuration together, in the presumably more common case when special firewall rules are not needed.

Lassulus replied:
ok, but it's very hard to have a different port configured and not open the ports on all interfaces then. I would prefer it, if this would be left out of the module.

gazally replied:
I removed it.

Lassulus commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -0,0 +1,119 @@
+      config = mkOption {
+        type = str;
+        default = "";
+        description = ''
```
would an example config would be appropriate? or is it to long to be inline here? maybe a minimal/partial one

gazally replied:
What do you think of adding to the description a suggestion to use `nix-shell -p yggdrasil --run "yggdrasil -genconf"` to see what the available options are?

Lassulus commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -0,0 +1,119 @@
+      wantedBy = [ "multi-user.target" ];
+
+      preStart = ''
+        mkdir -p /var/lib/yggdrasil/
```
there is `systemd.tmpfiles` which is now the preferred way to setup the working directory of services, you can see some implementations in this PR:
https://github.com/NixOS/nixpkgs/pull/56265/files

Lassulus replied:
also, maybe we want to run the service as another user than root? IMHO it would be preferable to run services as less privileged users.

gazally replied:
Both of those are very good suggestions and I'll work on incorporating them.

aanderse replied:
In this case you want `StateDirectory` instead of `tmpfiles`. Only use `tmpfiles` when `StateDirectory`, `RuntimeDirectory`, `CacheDirectory`, and `LogsDirectory` aren't flexible enough.

aanderse replied:
Would `DynamicUser` be sufficient for this module?

gazally replied:
I think so and I'm working on trying it out.  One snag which I have run into is that the admin utility `yggdrasilctl` has a hardcoded expectation to either find its configuration file in `/etc/yggdrasil.conf` or a unix socket at `/var/run/yggdrasil.sock` and it looks like `DynamicUser` only lets services have write access to subdirectories of `/etc` and `/run`.  Is that correct?  If so I will have to patch `yggdrasilctl` to change its default filenames.

aanderse replied:
`/var/run` is a symlink to `/run`. I think most distros have adopted that.

You could symlink the configuration file to `/etc` via `environment.etc` and if your user has the proper permissions I believe that would work, or run a patch in the source code to update the socket directory.

I'd also like to mention there is no hard requirement to use `DynamicUser` user here. Just use it when it makes sense to.

gazally replied:
Thanks for the symlink suggestion.

I've been unable to get yggdrasil to run using `DynamicUser` or any user other than root, even after experimenting with `capsh` to try to give that user as many of root's capabilities as possible.  So it is still running as root, but with capabilities stripped and many of systemd's sandboxing features applied, including most which are suggested in #20186.

gazally replied:
Just figured out DynamicUser and pushed an update.

Lassulus commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -0,0 +1,119 @@
+        '');
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/yggdrasil -useconffile /var/lib/yggdrasil/yggdrasil.conf";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
```
In packages pkgs. is normally not referenced via pkgs. instead we import the needed dependency directly.
So in this case, we would remove the pkgs from the arguments in line 1 and change it to coretutils

Lassulus replied:
ah nvm, my search confused me and I ended up here :D

Lassulus commented on pkgs/tools/networking/yggdrasil/default.nix:
```diff
@@ -0,0 +1,36 @@
+{ pkgs, lib, buildGoModule, fetchFromGitHub }:
```
```suggestion
{ lib, buildGoModule, fetchFromGitHub }:
```

gazally replied:
Fixed.

JohnAZoidberg commented:
Make sure to use the email address that is linked to your GitHub account for the commits, then they're linked to your profile account.

kalbasit commented on pkgs/tools/networking/yggdrasil/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+  modSha256 = "1r942mqlq9lqxizxi33768jp4j706mwa2fliz0qbkh98r4ri2292";
+
+  subPackages = [ "." ];
```
```suggestion
```

Lassulus replied:
this yields this error:
`contrib/config/yggdrasilconf.go:65:14: cannot use flag.Arg(1) (type string) as type []string in assignment`

gazally replied:
I copied and pasted that line from the buildGoModule example in the Nixpkgs manual, without any idea what it does (and the manual is not helpful on that point).  When I remove it the package builds without error.  NixOS 19.03 x86_64-linux, @Lassulus  what are you building this on?

Lassulus replied:
I got this when I also removed the custom buildPhase. It works though if I add only
```
  subPackages = [ "cmd/yggdrasil" ];
```
or
```
  subPackages = [ "cmd/yggdrasilctl" ];
```
but not when both or none are present

gazally replied:
Thanks for figuring out that it worked with only one thing in subPackages.  That allowed me to track the problem down to a bug in buildGoModule, which I just submitted as #63923.

kalbasit commented on pkgs/tools/networking/yggdrasil/default.nix:
```diff
@@ -0,0 +1,36 @@
+  buildPhase = ''
+    export PKGNAME="${pname}"
+    export PKGVER="${version}"
+    ./build
```
the only good thing this custom build is getting you is the ldflags. Please add the ldflags manually and remove both of your custom phases.

Example:

https://github.com/NixOS/nixpkgs/blob/576ffb1f32267a022efc0976c4b95b85b4b5d16d/pkgs/development/tools/conftest/default.nix#L14-L17

gazally replied:
Ok, I tried it without the custom phases and with this:
```

 buildFlagsArray = ''
     -ldflags=
         -X github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.buildVersion=${version}
         -X github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.buildName=${pname}
  '';
```
which is what it looks to me like yggdrasil's build script is setting the ldflags to, see [here](https://github.com/yggdrasil-network/yggdrasil-go/blob/master/build).

Result:
```
can't load package: package github.com/yggdrasil-network/yggdrasil-go: unknown import path "github.com/yggdrasil-network/yggdrasil-go": cannot find module providing package github.com/yggdrasil-network/yggdrasil-go
builder for '/nix/store/4d1c005ypmly2vx45apn733jcs7f820s-yggdrasil-0.3.5.drv' failed with exit code 1
error: build of '/nix/store/4d1c005ypmly2vx45apn733jcs7f820s-yggdrasil-0.3.5.drv' failed
```

gazally replied:
Hi @kalbasit, the problem was that setting subPackages to `["."]` was not the right thing to do, and there is a bug in buildGoModule which prevents the derivation from building when you set subPackages to `["cmd/yggdrasil" "cmd/yggdrasilctl"]`.  I just submitted #63923 regarding the bug.

kalbasit commented on pkgs/tools/networking/yggdrasil/default.nix:
```diff
@@ -0,0 +1,36 @@
+    homepage = "https://yggdrasil-network.github.io/";
+    license = licenses.lgpl3;
+    maintainers = with maintainers; [ gazally ];
+    platforms = platforms.linux;
```
```suggestion
```

gazally replied:
I've just force pushed a commit which I think addresses all your issues with the package.  Please take a look.  I'm still working on the service module.

aanderse commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -0,0 +1,119 @@
+        chmod 700 /var/lib/yggdrasil
+        chown -R root:root /var/lib/yggdrasil
+        '' + (if (cfg.configFile == null) then ''
+           ${cfg.package}/bin/yggdrasil -genconf > /var/lib/yggdrasil/yggdrasil.conf
```
Does the `yggdrasil` binary ever modify the `.conf` file?

gazally replied:
No, it doesn't.

aanderse commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -0,0 +1,119 @@
+        '' + (if (cfg.configFile == null) then ''
+           ${cfg.package}/bin/yggdrasil -genconf > /var/lib/yggdrasil/yggdrasil.conf
+        '' else ''
+          ${cfg.package}/bin/yggdrasil -normaliseconf -useconffile \
```
Is the output of this command dependant on the system environment, or is this something that could be run beforehand when the system configuration is being built?

gazally replied:
`cfg.configFile` could be a file somewhere in the system environment (such as in `/run/keys`, delivered there by `nixops send-keys`.

aanderse replied:
I understand now. Thanks for explaining.

You might consider if `RuntimeDirectory` is a better place to write the config file to because the contents of that directory are cleaned up after the services stops.

gazally replied:
The config file is now written to `RuntimeDirectory` and accessed from `/etc` via a symlink.

gazally commented:
I just pushed a new version which applies many of systemd's sandboxing features to the yggdrasil service. I've got this running on 4 computers in 2 locations including one Raspberry Pi (aarch64) and have verified that local multicast peering and remote peering both work.

aanderse commented:
Great work! Any chance for a NixOS test?

gazally commented:
> Any chance for a NixOS test?

I also think it's a good idea, so I'll find myself a Perl tutorial and see what I can do.

Also, I've decided that it would be better to have `service.yggdrasil.config` be a Nix attribute set, and then use `builtins.toJSON` to convert it into a config file for yggdrasil, so I'll be making that change as well unless anyone objects.

gazally commented:
In this latest push, I have:

- Added a test.
- Made `services.yggdrasil.config` into a Nix attribute set.
- Added a new option `services.yggdrasil.denyDhcpcdInterfaces` so dhcpcd broadcasts can be suppressed on TAP interfaces.
- Added an assertion on `config.networking.enableIPv6`.

arcnmx commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -0,0 +1,181 @@
+      wantedBy = [ "multi-user.target" ];
+
+      preStart = ''
+        ${generateConfig} | yggdrasil -normaliseconf -useconf > /run/yggdrasil/yggdrasil.conf
```
```suggestion
        ${generateConfig} > $RUNTIME_DIRECTORY/yggdrasil.conf
```
`normaliseconf` is a no-op since yggdrasil populates the config with defaults when it reads it regardless, so a `preStart` really isn't necessary except in the case where you need to do the jq config merge (or genconf).

gazally replied:
My intention with the use of `/etc` and `-normaliseconf` was to make it possible for someone armed with sudo to look in the obvious place (`/etc/yggdrasil.conf`) and find all the configuration values that their yggdrasil service is using. It's not necessary to get the service to work, but I found it helpful when experimenting with merging private keys from a file with the rest of the configuration in Nix.

gazally replied:
yggdrasilctl looks first for `/etc/yggdrasil.conf` and if that is not found then it looks for `/var/run/yggdrasil/yggdrasil.sock`. If you don't have either one of those on your system or if you are running multiple instances of yggdrasil then you can use the `-endpoint` argument to yggdrasilctl to tell it where the socket is.

arcnmx commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -0,0 +1,181 @@
+      '';
+
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/yggdrasil -useconffile /etc/yggdrasil.conf";
```
```suggestion
        ExecStart = "${cfg.package}/bin/yggdrasil -useconffile \${RUNTIME_DIRECTORY}/yggdrasil.conf";
```
I think? Though this can just refer to the config path directly unless it needs to be merged by preStart.

arcnmx commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -0,0 +1,181 @@
+        RuntimeDirectory = "yggdrasil";
+        RuntimeDirectoryMode = "0700";
+        BindReadOnlyPaths = mkIf configFileProvided
+          [ "${cfg.configFile}:/run/yggdrasil/configFile.json" ];
```
```suggestion
          [ "${cfg.configFile}:/tmp/configFile.json" ];
```
Since I don't think you can do variable interpolation here, this seems like a more flexible path to use? Using private /tmp also lets one expose the RuntimeDirectory with more slightly liberal permissions (to say, share the admin socket to allow a group to use yggdrasilctl) without exposing the key config data.

gazally replied:
I don't think `DynamicUser=true` is compatible with allowing group access to the socket, since the group id might be different on successive restarts of the service.  But access could be granted to a group to run yggdrasilctl via a sudoers rule without exposing the config data (to my knowledge yggdrasilctl doesn't have a command to reveal the private keys).

arcnmx replied:
> I don't think DynamicUser=true is compatible with allowing group access to the socket, since the group id might be different on successive restarts of the service

Yeah to be fair it is awkward to fix up those permissions, that's what socket units are for more than anything else.

> access could be granted to a group to run yggdrasilctl via a sudoers rule

Eh that's... generally not a good idea for the same reason setuid can go horribly wrong when used with programs that aren't designed to be secure. You could probably make yggdrasilctl do something silly like read out /etc/shadow for you or something.

arcnmx commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -0,0 +1,181 @@
+    networking.firewall.allowedUDPPorts = mkIf cfg.openMulticastPort [ 9001 ];
+
+    # Make yggdrasilctl available on the command line.
+    environment.systemPackages = [ cfg.package ];
```
Okay so this one's just a personal wish and probably too opinionated but it'd be neat if we could get this to produce multiple outputs somehow, considering only `yggdrasilctl` really needs to be in the environment. They could also just be two separate derivations sharing the same source, not sure what's typical/standard for nixpkgs...

gazally replied:
One reason to have yggdrasil in the environment is that `yggdrasil -genconf` and `yggdrasil -genconf -json` are useful commands to have available when you are working on your yggdrasil configuration.

arcnmx replied:
Yup that's the main reason to run it from outside of the service, though that seems more what `nix run nixpkgs.yggdrasil` is for rather than making a one-time command a permanent part of the user environment - also ideally the nixos module documents all the settings so you wouldn't even have to do that!

arcnmx reviewed (commented):
Nice! I'm looking forward to using this to replace my current ad-hoc [package](https://github.com/arcnmx/nixexprs/blob/master/pkgs/public/yggdrasil/default.nix) and [module](https://github.com/arcnmx/nixexprs/blob/master/modules/nixos/yggdrasil.nix). Leaving a few comments aiming to avoid preStart where it's unnecessary and keep the paths used a bit less hard-coded, mostly just because I'm often interested in running multiple yggdrasil services at once.

gazally commented on pkgs/tools/networking/yggdrasil/default.nix:
```diff
@@ -0,0 +1,36 @@
+    sha256 = "0cbj9hqrgn93jlybf3mfpffb68yyizxlvrsh1s5q21jv2lhhjcwj";
+  };
+
+  modSha256 = "1r942mqlq9lqxizxi33768jp4j706mwa2fliz0qbkh98r4ri2292";
```
Yes, something changed.  I've updated it and will push a branch rebased on master shortly.

gazally commented on pkgs/tools/networking/yggdrasil/default.nix:
```diff
@@ -0,0 +1,36 @@
+     -ldflags=
+         -X github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.buildVersion=${version}
+         -X github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.buildName=${pname}
+         -s -w
```
@kalbasit I think this is a question for you.

I put those flags in `ldflags` to make it match what [yggdrasil's build script](https://github.com/yggdrasil-network/yggdrasil-go/blob/master/build) does.   I'm also a go build system newbie so I don't know if the way they have set up their build is unusual or not.

Lassulus commented:
so whats the state of this? should we merge this now and work on the multiple instance stuff later? or does someone wants to implement it now?

aanderse commented:
Would the multiple instance stuff cause breaking api changes?

arcnmx commented:
I wouldn't bother treating my comments as anything other than discussion points; I'm particularly interested due to how I use yggdrasil but that's not really relevant - and I can of course always continue to use a stripped-down module. The main areas I pointed out were paths that were a bit strangely hard-coded or inefficient/unnecessary, but I don't have a good understanding of nixos conventions or style regarding whether any of that actually matters (e.g. is polluting /etc bad or do plenty of other modules do it too?)

Also, what was the reason for meta.platforms being removed?

gazally commented:
Upstream released 0.3.6 last week and I am hoping to find time this week to update the package.

> I don't have a good understanding of nixos conventions or style regarding whether any of that actually matters (e.g. is polluting /etc bad or do plenty of other modules do it too?)

I'm in the same boat.  I observed that other NixOS modules put things in /etc and did so too.

> Also, what was the reason for meta.platforms being removed?

That was an accident.  Thanks for catching it.

gazally commented:
yggdrasil 0.3.5 and 0.3.6 now do not build due to a broken dependency, and upstream has marked it wontfix. They're recommending use of their develop branch or waiting for 0.3.7.  Before that happened I was running into #64346 while trying to test the package on my mix of machines running 19.03 and unstable, so I guess I'll wait a bit.

gazally committed:
nixos/yggdrasil: remove DynamicUser for systemd 243

gazally commented:
I've just pushed an update of the package to the latest version, 0.3.8. There is now an additional commit which removes DynamicUser from the service module.  Yggdrasil works with DynamicUser set on systemd 242, but fails to open its network adapter if DynamicUser is set on systemd 243, and systemd 243 is now in Nixpkgs unstable. I have reported this issue to yggdrasil upstream.

gazally commented:
yggdrasil 0.3.8 does not build on Go 1.13, which is pull request #68135.  This will be fixed in yggdrasil 0.3.9.

Lassulus commented:
soo, I guess we can merge this now and fix the building on go1.13 later?

aanderse commented on maintainers/maintainer-list.nix:
```diff
@@ -2350,6 +2350,11 @@
     github = "gavinrogers";
     name = "Gavin Rogers";
   };
+  gazally = {
```
You need to add your github id to this as part of recent requirements.

ehmry commented:
Merged to master, but `githubId` is still missing.
